source/common/stats/stats_impl.h

```diff
@@ -473,5 +477,131 @@ class IsolatedStoreImpl : public Store {
   IsolatedStatsCache<Histogram, HistogramImpl> histograms_;
 };
 
+/**
+ * Structure to hold statistical data of histogram.
+ */
+struct HistogramStatistics {
+public:
+  HistogramStatistics(const std::string name, histogram_t* histogram_ptr) : name_(name) {
+    double arr[9] = {0, 0.25, 0.5, 0.75, 0.90, 0.95, 0.99, 0.999, 1};
+    double out[9];
+    double* arrptr = arr;
+    double* outptr = out;
+    hist_approx_quantile(histogram_ptr, arrptr, 9, outptr);
+    p0_ = out[0];
+    p25_ = out[1];
+    p50_ = out[2];
+    p75_ = out[3];
+    p90_ = out[4];
+    p95_ = out[5];
+    p99_ = out[6];
+    p999_ = out[7];
+    p100_ = out[8];
+  }
+
+  std::string name() const { return name_; }
+
+  std::string summary() const {
+    return fmt::format("P0: {} , P25: {}, P50: {}, P75: {}, P90: {}, P95: {}, P99: {}, P100: {}",
+                       p0_, p25_, p50_, p75_, p90_, p95_, p99_, p999_, p100_);
+  }
+
+private:
+  const std::string name_;
+  double p0_;
+  double p25_;
+  double p50_;
+  double p75_;
+  double p90_;
+  double p95_;
+  double p99_;
+  double p999_;
+  double p100_;
+};
+
+/**
+ * BaseSink implementation for all Sinks. This mainly  .
+ */
+class BaseSink : public Sink {
+public:
+  void onHistogramComplete(const Histogram& hist, uint64_t value) override {
+    ThreadLocalHistograms tls_histograms = hist_tls_->getTyped<ThreadLocalHistograms>();
+    auto tls_hist_iterator = tls_histograms.histograms_.find(hist.name());
+    if (tls_hist_iterator != tls_histograms.histograms_.end()) {
+      hist_insert(tls_hist_iterator->second, value, 1);
```

Comment:
value is a uint64_t.  Not sure it needs that higher big, but there is much faster insert method that doesn't require conversion into floating point.

`hist_insert_intscale`, just set the scale to 0.  If you want to handle the range between int64_t and uint64_t, you can say something like:

```
int scale = 0;
int64_t signed_scaled_value;
if (value > MAX_INT64) { signed_scaled_value = value / 10; scale = 1; }
else signed_scaled_value = value;
hist_insert_intscale(tls_hist_iterator->second, signed_scaled_value, scale, 1);
```

Avoiding the floating point math can cut as much as 70% off the runtime of the insert.

Reply:
@postwait Thanks for your review. I will go through it and see if we can use the faster hist_insert_intscale method. Do you think we should experiment with any different configs (I do not know what can tweaked)  to get the better histogram results or defaults are good enough?

Reply:
+1 please avoid floating point. Everything should use the intscale variants per @postwait